genwinsyscallproto/genwinsyscalldef.go

```diff
@@ -0,0 +1,201 @@
+	if p.Flags&flags.ParamAttributes_In != 0 {
+		w.WriteString("/*in*/")
+	}
+	if p.Flags&flags.ParamAttributes_Out != 0 {
+		w.WriteString("/*out*/")
+	}
+	if p.Flags&flags.ParamAttributes_Optional != 0 {
+		w.WriteString("/*optional*/")
+	}
+	if p.Flags&flags.ParamAttributes_HasDefault != 0 {
+		w.WriteString("/*hasdefault*/")
+	}
+	if p.Flags&flags.ParamAttributes_HasFieldMarshal != 0 {
+		w.WriteString("/*hasfieldmarshal*/")
+	}
+	if p.Flags&flags.ParamAttributes_Unused != 0 {
+		w.WriteString("/*unused*/")
+	}
+	w.WriteString(p.Name.String())
+}
+
+func writeType(b io.StringWriter, f *winmd.Metadata, p winmd.Type) error {
+	// Special case: *void is unsafe.Pointer
+	if p.Kind == flags.ElementType_PTR {
+		if t, ok := p.Value.(winmd.Type); ok {
+			if t.Kind == flags.ElementType_VOID {
+				b.WriteString("unsafe.Pointer")
+				return nil
+			}
+		}
+	}
+
+	switch p.Kind {
+	// Translate ECMA-335 primitive types to Go types.
+	case flags.ElementType_BOOLEAN:
+		b.WriteString("bool")
+	case flags.ElementType_CHAR:
+		// TODO: Is there a better representation of CHAR?
+		b.WriteString("uint16")
+	case flags.ElementType_I1:
+		b.WriteString("int8")
+	case flags.ElementType_U1:
+		b.WriteString("uint8")
+	case flags.ElementType_I2:
+		b.WriteString("int16")
+	case flags.ElementType_U2:
+		b.WriteString("uint16")
+	case flags.ElementType_I4:
+		b.WriteString("int32")
+	case flags.ElementType_U4:
+		b.WriteString("uint32")
+	case flags.ElementType_I8:
+		b.WriteString("int64")
+	case flags.ElementType_U8:
+		b.WriteString("uint64")
+	case flags.ElementType_R4:
+		b.WriteString("float32")
+	case flags.ElementType_R8:
+		b.WriteString("float64")
+
+	// ECMA-335 distinguishes uintptr and intptr, Go only has uintptr used in both cases.
+	case flags.ElementType_I, flags.ElementType_U:
+		b.WriteString("uintptr")
+
+	case flags.ElementType_VOID:
+		// We catch "*void" with a special case above. We should never see simply VOID.
+		return errors.New("unexpected primitive type: VOID")
+
+	// If this is not a simple value type, there will be p.Value. Handle all those cases here.
+	default:
+		if p.Kind == flags.ElementType_PTR {
+			b.WriteString("*")
+		}
+		return writeTypeValue(b, f, p.Value)
+	}
+	return nil
+}
+
+func writeTypeValue(b io.StringWriter, f *winmd.Metadata, value any) error {
+	switch v := value.(type) {
+	case winmd.CodedIndex:
+		switch v.Tag {
+		case coded.TypeDefOrRefOrSpec_TypeDef:
+			record, err := f.Tables.TypeDef.Record(v.Index)
+			if err != nil {
+				return err
+			}
+			b.WriteString(record.Name.String())
+
+		case coded.TypeDefOrRefOrSpec_TypeRef:
+			record, err := f.Tables.TypeRef.Record(v.Index)
+			if err != nil {
+				return err
+			}
+			b.WriteString(record.Name.String())
+
+		default:
+			return fmt.Errorf("unexpected coded index value: %#v", v)
+		}
+
+	// Types can nest. A pointer to another type is a very common case.
+	case winmd.Type:
+		return writeType(b, f, v)
```

Comment:
I'm struggling to actually create a cycle here because this code only handles `winmd.Type`, not `*winmd.Type`. (I set up a map and I'm trying to write a test case.) `any` makes this kinda confusing because internally it points to a copy of the data, but it can't be modified (as far as I can tell) to actually establish a cycle by re-pointing it an existing type in memory.

Reply:
I've given another thought at this and I think that you are right, we can't have cycles the way it is implemented. Nothing references a `winmd.SigType` nor `winmd.SigArray`, and those are the only signature types that recurse, so we are safe for now 😄 .